### DISCUSSION TITLE
Invalid Url when validating a MSSQLServer connection string

### DISCUSSION THREAD

**Discussion by avenmia** [colinhacks/zod#2031] [Q&A] [answered]
I'm using zod with the t3 stack. 

The server runs the following validation on schemas.
```ts
export const serverSchema = z.object({
  DATABASE_URL: z.string().url(),
  NODE_ENV: z.enum(["development", "test", "production"]),
  NEXTAUTH_SECRET:
    process.env.NODE_ENV === "production"
      ? z.string().min(1)
      : z.string().min(1).optional(),
  NEXTAUTH_URL: z.preprocess(
    // This makes Vercel deployments not fail if you don't set NEXTAUTH_URL
    // Since NextAuth.js automatically uses the VERCEL_URL if present.
    (str) => process.env.VERCEL_URL ?? str,
    // VERCEL_URL doesn't include `https` so it cant be validated as a URL
    process.env.VERCEL ? z.string() : z.string().url(),
  ),
  CLIENT_ID: z.string(),
});
```

My database url looks like the following:
```ts
DATABASE_URL="sqlserver://localhost:1433;initialCatalog={MyDatabase};integratedSecurity=true;trustServerCertificate=true;"
```

This is the error that's occurring: 
❌ Invalid environment variables:
 DATABASE_URL: Invalid url

I see the code is using the new URL constructor and when I plug it in to dev tools, I get the following which may cause the error:
![image](https://user-images.githubusercontent.com/17712276/218238469-a5eaa744-3f84-4bf9-9155-2d34f2e2e009.png)

Is this intentional and should I be using some other form of validation or is this a bug?

Thank you

**Comment by JacobWeisenburger** [comment, marked as answer]
This is not a Zod bug. This is the way JS URL works.

```ts
const DATABASE_URL = "sqlserver://localhost:1433;initialCatalog={MyDatabase};integratedSecurity=true;trustServerCertificate=true;"

try {
    new URL( DATABASE_URL )
} catch ( error ) {
    console.error( error.message )
    // Invalid URL: 'sqlserver://localhost:1433;initialCatalog={MyDatabase};integratedSecurity=true;trustServerCertificate=true;'
}
```

Perhaps you can use `z.string().regex( /some regex/ )`.

---

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by avenmia** [reply]
That makes sense, thank you
